classes/statistics/extStatistics.sc

```diff
@@ -480,6 +480,91 @@ oscorr(x,y);
 		};
 		^p
 	}
+
+	// Simple Linear Regression (Least Squares) - Linear Fit
+	linearFit {
+		var tmpThis;
+		var size, x, y;
+		var xMean, yMean, sSxy, sSxx, b0, b1;
+		var errStr = "Support is only provided for 2D line fitting. Supply y or x, y pairs only.";
+		var err = Error.new(errStr);
+
+		size = this.size;
+		this.rank.switch(
```

Comment:
I find this cases a bit loaded with implicit assumptions.
Wouldn't it perhaps be better to make a method that takes one argument, thus instead of
```[[a1, a2, a3], [b1, b2, 3]].linearFit```
```linearFit([a1, a2, a3], [b1, b2, b3])```?
Then you can explicitly leave away the argument:
```linearFit([a1, a2, a3])``` and get a series [1, 2, 3] for the second arg.

Reply:
From the help, the initially proposed use pattern is:

```
// If this is a rank 1 collection, intrepret these values as y and the indices as x:
(
var size = 25;
var y = Array.fill(size, { size.rand }).sort;
var coeffs = y.linearFit;

y.plot("random vals", discrete: true, minval: 0.0, maxval: size);
Array.series(size, coeffs.at(0), coeffs.at(1)).plot("linear fit", minval: 0.0, maxval: size)
)
```

and

```
// If this is a rank 2 collection, intrepret these values as x, y pairs:
(
var size = 25;
var x = Array.fill(size, { size.rand }).sort;
var y = Array.fill(size, { size.rand }).sort;
var xy = Array.with(x, y).flop;
var coeffs = xy.linearFit;

Array.series(size, coeffs.at(0), coeffs.at(1)).plot("linear fit", minval: 0.0, maxval: size)
)

```

In other words, I'm expecting interleaved xy pairs or just y.

----
So are you suggesting a better interface would look like: `y.linearFit` and `y.linearFit(x)`?

As the second form would also look like `linearFit(y, x)`, I think I'd find this confusing....

Reply:
From my experience, interpreting the elements of an array differently depending on its internal structure is not so good. Instead of making the code document itself, it requires a somewhat hard to understand documentation. I would see it as a good reason to do this anyway if the arrays can be of any dimension and if the structure keeps the same meaning. But both seem not to be the case here.

I need to understand the thinking behind this first, so here some questions (sorry!):
- Is there a particular use case where it comes handy to have each value as a pair? Do you think them as points?
- in the case of rank one, do you want the values to be interpreted as first elements of a pair each?
- Is there a reason why one may want to extend the data dimensions?

Reply:
> - Is there a particular use case where it comes handy to have each value as a pair? Do you think them as points?

Yes! Having pairs (points) is immensely useful for unevenly sampled data.

When we have sparsely sampled data, or critically reduced data, we need to know both x, the sample index (time for my application), and y, the measured value.

Here's one of my immediate use cases:

1. measuring instantaneous signal power (discrete, evenly sampled time)
2. reduce to a sparse set of relevant peaks (no longer evenly sampled in time)
3. fit sparse set of x,y points to estimate trend

> - in the case of rank one, do you want the values to be interpreted as first elements of a pair each?

My current thinking is no.

For rank two, we have xy points: [ [x0, y0], [x1, y1] ... ]

Where x is the measurement sample index. If unevenly sampled, we definitely need x.

For rank one, we just need the measurements, the y. We can assume that x is the array index, so we don't need to supply it.

I'm willing to be talked into an interface that looks like: `y.linearFit(nil)` and `y.linearFit(x)`. 

For `y.linearFit(nil)` we just use the indices as x, which will probably be the most common use. Though, as mentioned earlier, the unevenly sampled case using functional syntax would look like `linearFit(y, x)`, which could be confusing. But, if the documentation makes it otherwise clear, I think that would be OK.

BTW, I've just had a very quick review of Numpy/SciPy interfaces, they are of the form `linearFitAlgo(x, y, ...)`, _always_ requiring x to be set. My preference is avoid supplying x if we don't need to.

> Is there a reason why one may want to extend the data dimensions?

In the abstract, yes. But, I wasn't planning to go there as a generalized solution takes rather more effort. (As much as I'd love to have all the fitting found in Numpy/SciPy, that's not exactly a feasible target.) My goal here is just a simple 2D line fit, as this suits my actual practical signal estimation problem... and can be used for all sorts of useful tasks.

_This seems to suggest_ we shouldn't bother with arrays of different ranks, and refactoring to `y.linearFit(nil)` and `y.linearFit(x)` might be the way to go.

How do you feel about that?

Reply:
Thank you, this makes it very understandable. I find there are arguments for both possibilities. What tips me in the direction of `y.linearFit(x)` is that then you don't need a `flop` in the method, which is more efficient. In some examples, you currently have one flop outside and then one inside the call.

> My preference is avoid supplying x if we don't need to.

yes, I agree that is much better!

Reply:
> But just to make sure, because this is the decisive point: I still wonder if it might be relatively easy to add higher-dimensional points (perhaps later)? Some of the measures seem to generalize easily, but of course I don't know what I may be missing.

I would say there are actually two issues here:

1. [least squares polynomial fitting](https://mathworld.wolfram.com/LeastSquaresFittingPolynomial.html)
2. multidimensional / multivariate fitting

My modest goal has been to target least squares linear (degree 1) fitting. Particularly as the solution is fairly easy.

Generalizing LS to higher degrees is well defined, and possible given the Matrix class. But... given the variety of implementations (and practical approaches) in the wild, I'm thinking like the higher degree problem is something I'm not able to get into at the moment.

As for making the problem multivariate, there seem to be a number of approaches. But a comment from Bill Press ([Numerical Recipes Forum](http://numerical.recipes/forum/showthread.php?t=45)) summarizes:

> The trick described in 15.4 for turning any linear least square problem into one that can be solved with one-dimensional routines is, basically, to renumber all your multi-dimensional points into a single one-dimensional sequence (the order doesn't matter), and kludge up the user-supplied function to supply basis function values at these points.

Anyway... I'm feeling this second part of the problem (the implemented algorithms) is beyond me!

.. so yes.. I was just attempting to go for the most simple case, linear 2D fitting (and threw in the Theil-Sen Linear Regression, because it is a simple "robust" alternative).

Reply:
Ah yes. It is then good to just do the base case. 

Actually, I am open to both variants, it mainly depends on what the more typical use case is, i.e. having an array of pairs or a pair of arrays. If you mainly work with arrays of pairs, then the current version is fine. If not, then it may be better to switch to the simpler-to-read version that takes `x` as an argument.

Reply:
Let's go with `x` as an argument. That's the change I made here 09122d09609e6d8d68635d2df106907cfd748076.

I think most uses will just use the auto-index form. The pairs approach is useful for sparse data. Given the change in the interface, the use pattern should be very clear.

(For me, I just have to collect / reshape my points as need be, anyway!)

Everything else look ok?